File: 2709. Greatest Common Divisor Traversal/canTraverseAllPairs.py
class UnionFind:
  def __init__(self, n) -> None:
    self.par = [i for i in range(n)]
    self.size = [1] * n
    self.count = n

  def find(self, x):
    if self.par[x] != x:
      self.par[x] = self.find(self.par[x])
    return self.par[x]
  
  def union(self, x, y):
    px, py = self.find(x), self.find(y)
    if px == py:
      return
    
    if self.size[px] < self.size[py]:
      self.par[px] = py
      self.size[py] += self.size[px]
    else:
      self.par[py] = px
      self.size[px] += self.size[py]
    self.count -= 1

class Solution:
  def canTraverseAllPairs(self, nums: list[int]) -> bool:
    uf = UnionFind(len(nums))

    factorIndex = {}
    for i, n in enumerate(nums):
      f = 2
      while f * f <= n:
        print('N: ', f, n)
        if n % f == 0:
          if f in factorIndex:
            uf.union(i, factorIndex[f])
          else:
            factorIndex[f] = i
          while n % f == 0:
            n = n // f
        f += 1

      if n > 1:
        if n in factorIndex:
          uf.union(i, factorIndex[n])
        else:
          factorIndex[n] = i

    print('uf.count: ', uf.count)
    return uf.count == 1

solution = Solution()
nums = [3, 9, 5]
print("canTraverseAllPairs: ", solution.canTraverseAllPairs(nums))

File: 2709. Greatest Common Divisor Traversal/canTraverseAllPairs.js

class UnionFind {
  constructor(n) {
    this.parent = [];
    for (let i = 0; i < n; i++) {
      this.parent.push(i);
    }
    this.sizes = Array(n).fill(1);
    this.count = n;
  }

  find(x) {
    if (this.parent[x] !== x) {
      this.parent[x] = this.find(this.parent[x]);
    }

    return self.parent[x];
  }

  union(x, y) {
    const rootX = this.find(x);
    const rootY = this.find(y);

    if (rootX === rootY) {
      return;
    }

    if (this.sizes[rootX] < this.sizes[rootY]) {
      this.parent[rootX] = rootY;
      this.sizes[rootY] += this.sizes[rootX];
    } else {
      this.parent[rootY] = rootX;
      this.sizes[rootX] += this.sizes[rootY];
    }
    this.count--;
  }
}

/**
 * @parentam {number[]} nums
 * @return {boolean}
 */
const canTraverseAllPairs = (nums) => {

};